### PR TITLE
[scroll-animations-1] Use CSSTypedOM instead of strings (#5213)

### LINKED ISSUE #5213: [scroll-animations-1] Use of DOMString in API surface
Small point that came up in the [TAG Review](https://github.com/w3ctag/design-reviews/issues/521), the [ScrollTimeline](https://drafts.csswg.org/scroll-animations-1/#scrolltimeline-interface) interface is currently using DOMString to represent keyword and length/percentage values. This should use the appropriate [CSS Typed OM](https://drafts.css-houdini.org/css-typed-om/) classes instead.

Using strings for CSS values is a historical artifact we're trying to eliminate, new APIs shouldn't use them.

### PR DESCRIPTION
**Pull request by majido**
Use appropriate CSSTypedOM instead of DOMStrings in the API. Fixes #5213.

Changes:
 - For offsets use `CSSNumericValue` or `CSSKeywordish` instead of DOMString
 - When setting the offset, the grammars for these values is check to ensure only the valid subset can be set.
 - When computing effective scroll offset check the offset actual type and interpret it accordingly
 - Update examples to showcase the API and add noted to clarify the usage.

### CODE REVIEW

File: scroll-animations-1/Overview.bs [hunk 1]
```diff
@@ -487,21 +492,37 @@ The procedure to <dfn>resolve a container-based offset</dfn> given
...
     <div class="switch">
 
-    :   <var>offset</var> is <code>auto</code>
+    :   If <var>offset</var> is <code>auto</code>:
     ::  Either the beginning or the ending of {{source}}'s scroll range in
         {{orientation}} depending on whether the offset is {{start}} or {{end}}.
 
-    :   <var>offset</var> is a <<length-percentage>>
+    :   If <var>offset</var> is a {{CSSNumberish}} and matches <<number>>:
+    ::  The distance indicated by the value in <a value>pixel unit</a> along
```
Comment (andruud) [inline scroll-animations-1/Overview.bs:500]: I was under the impression that treating plain numbers as `px` was a pattern we didn't want? Maybe get someone else's opinion on this. (@tabatkins ?)

File: scroll-animations-1/Overview.bs [hunk 2]
```diff
@@ -487,21 +492,37 @@ The procedure to <dfn>resolve a container-based offset</dfn> given
...
+        {{source}}'s scroll range in {{orientation}}.
+
+
+    :   If <var>offset</var> is a {{CSSNumberish}} and matches
```
Comment (andruud) [inline scroll-animations-1/Overview.bs:504]: Can we link "matches" to https://drafts.css-houdini.org/css-typed-om-1/#cssstylevalue-match-a-grammar ?

File: scroll-animations-1/Overview.bs [hunk 3]
```diff
@@ -348,11 +354,14 @@ behavior under different directionalities and writing modes.
 <pre class="idl">
 enum ScrollTimelineAutoKeyword { "auto" };
 
+typedef (CSSNumberish or ScrollTimelineAutoKeyword) ContainerBasedOffset;
```
Comment (andruud) [inline scroll-animations-1/Overview.bs:357]: We should IMO also accept `CSSKeywordValue` matching `auto`.

Would you want to use `CSSKeywordish` here (if it existed)?

Comment (andruud) [reply scroll-animations-1/Overview.bs:357]: My bad, CSSKeywordish already exists: https://drafts.css-houdini.org/css-typed-om-1/#typedefdef-csskeywordish

Do we want to use that somehow, then?

File: scroll-animations-1/Overview.bs (hunk 1 above)
Comment (majido) [reply scroll-animations-1/Overview.bs:500]: I personally felt allowing numbers to be interpreted as `px` would make the API easier to use in Javascript. Specially as it matches all legacy scroll related APIs where numbers are treated as pixels


But I can also see at least two possible concerns:
 - Automatic type conversion can be a footgun but perhaps this is limited enough to be safe
 - Having every downstream spec convert `<numbers>` differently would be bad. So perhaps have a common way to do this i.e., `<number>` -> `<length>` with pixel for all JS API's that accept length but want to accept number too.

BTW, does TypedOM have a special handling for `0` length for which in CSS the unit is optional?

In any case, this is a trade-off and @tabatkins expert's opinion is appreciated. :)

Comment (plinss) [reply scroll-animations-1/Overview.bs:500]: We really don't want to interpret bare numbers as px. The definition of CSSNumberish interprets numeric values as a CSS "number" not a length. `offset` should simply be a CSSNumericValue or a CSSStyleValue.

File: scroll-animations-1/Overview.bs (hunk 3 above)
Comment (majido) [reply scroll-animations-1/Overview.bs]: Done. As part of this I now have a procedure that rectifies strings into CSSKeyword and rejects invalid values at the time of setting the value.

File: scroll-animations-1/Overview.bs (hunk 2 above)
Comment (majido) [reply scroll-animations-1/Overview.bs]: Done.

File: scroll-animations-1/Overview.bs (hunk 1 above)
Comment (majido) [reply scroll-animations-1/Overview.bs]: Done. I removed CSSNumberish and just use CSSNumericValue which I think is the right type.

Also added a procedure to validate the incoming values when they are being set.

**Pull request merged**